bump to 24.1 - Single-container setup

jyotipm29 opened the pull request:
Upgrades:
  - Base Ubuntu Image: Upgraded from version 18.04 to 22.04
  - Galaxy: Upgraded from version 20.09 to 24.1
  - PostgreSQL: Upgraded from version 11 to 15
  - Python3: Upgraded from version 3.7 to 3.10 (Python 3.10 is set as the default interpreter)

Updates:
  - New Service Support:
    - Gunicorn: Replaces uWSGI as the web server for Galaxy. Installed by default inside Galaxy's virtual environment. Configured Nginx to proxy Gunicorn enabled on port 4001.
    - Celery: Installed by default inside Galaxy's virtual environment. Enabled Celery for distributed task queues and Celery Beat for periodic task running. RabbitMQ serves as the broker for Celery (if RabbitMQ is disabled, it defaults to PostgreSQL database connection). Redis is used as the backend for Celery (if Redis is disabled, it defaults to a SQLite database). Flower service is added for monitoring and debugging Celery.
    - RabbitMQ Management: Enabled the RabbitMQ management plugin on port 15672 for managing and monitoring the RabbitMQ server. The dashboard is exposed via Nginx and is accessible at the `/rabbitmq/` path. The default access credentials are `admin:admin`.
    - Redis: Added Redis server on port 6379 as a backend for Celery.
    - Flower: Added Flower service on port 5555 for monitoring and debugging Celery. The dashboard is exposed via Nginx and is available at the `/flower/` path. The default access credentials are `admin:admin`.
    - TUSd: Added TUSd server on port 1080 to support fault-tolerant uploads; Nginx is configured to proxy TUSd.
    - gx-it-proxy: Added gx-it-proxy service on port 4002 to support Interactive Tools.

  - Ansible Playbooks:
    - Added configure_rabbitmq_users.yml Ansible playbook, which removes the default `guest` user and adds `admin`, `galaxy`, and `flower` users for RabbitMQ during container startup.
  
  - Environment Variables:
    - Added `GUNICORN_WORKERS` and `CELERY_WORKERS` magic environment variables to set the number of Gunicorn and Celery workers, respectively, during container startup.
  
  - Configuration Changes:
    - Replaced the Galaxy Reports sample configuration file.
    - Removed galaxy_web, handlers, reports, and ie_proxy services from Supervisor.
    - Added Gravity for managing Galaxy services such as Gunicorn, Celery, gx-it-proxy, TUSd, reports, and handlers. It uses Supervisor as the process manager, with the configuration file located at `/etc/galaxy/gravity.yml`.
    - Added support for dynamic handlers (set as the default handler type).
    - Redis and Flower services are now managed by Supervisor.
    - Since Galaxy Interactive Environments are deprecated, they have been replaced by Interactive Tools (ITs). The sample configuration file `tools_conf_interactive.xml.sample` is placed inside `GALAXY_CONFIG_DIR`. Nginx is also configured to support both domain and path-based ITs.
    - Switched to using the `cvmfs-config.galaxyproject.org` repository for automatic configuration and updates of Galaxy project CVMFS repositories. Updated tool data table config path to include CVMFS locations from `data.galaxyproject.org` in `--privileged` mode.
    - Enabled IPv6 support in Nginx for ports 80 and 443.
    - Added Subject Alternative Name (SAN) extension (DNS:localhost and IP:127.0.0.1) while generating a self-signed SSL certificate.
    - Ensured the Nginx SSL certificate is trusted system-wide by adding it to the CA store.
    - Updated Galaxy extra dependencies.
    - Added `docker_net`, `docker_auto_rm`, and `docker_set_user` parameters for Docker-enabled job destinations.
    - Added `update_yaml_value.py` script to update nested key values in a YAML file.
    - Replaced `ie_proxy` with `gx-it-proxy`.
    - Replaced `nginx_upload_module` with `TUSd` for delegated uploads.
   - CI Tests:
      - Added dive tool for analyzing the docker image
      - Added test for check data persistence

bgruening commented:
Very cool, I triggered a test run. It would be nice if we can get tests to turn green at some point. But they probably also need to be updated.

Thanks a lot!

bgruening commented on .gitmodules:
```diff
@@ -1,7 +1,8 @@
 [submodule "galaxy/roles/galaxyproject.galaxyextras"]
 	path = galaxy/roles/galaxyprojectdotorg.galaxyextras
 	url = https://github.com/galaxyproject/ansible-galaxy-extras
-        branch = 20.05
```
currently you still need to change the URL to point to your fork

bgruening commented:
As a first step feel free to concentrate on the single-container option. Then the PR are easier to review.

jyotipm29 commented:
The single-container changes are done, and I’ve also updated the compose files. Everything seems to be working well, and tests in the forked repo have passed. I also updated the tool versions for the [workflow tests](https://github.com/jyotipm29/workflow-testing/commit/65a7eb958da079db964170c79b9ea00c2e91bc90), as the old ones seemed incompatible.

Next, I’ll work on adding the Rustus service, integrating interactive tools, and replacing Nginx with Traefik in the compose setup. Let me know if you have any feedback on the current changes. Thanks!

bgruening commented:
@jyotipm29 really impressive work. Thanks you a lot. 
Please split this PR into multiple PRs. So that we can get the single-container merged with documentation and you can try to hack further on the compose stuff.

mira-miracoli commented:
This looks really cool @jyotipm29 Thanks a lot! 🚀 
From what I see, the tests are failing due to github walltime limits and storage limits.
If you are sure that it works, maybe we could also test it manually again, we could already merge this. What do you think @sanjaysrikakulam ?
> Next, I’ll work on adding the Rustus service, integrating interactive tools, and replacing Nginx with Traefik in the compose setup. Let me know if you have any feedback on the current changes. Thanks!

A bit more important than the compose setup and traefik would be to successively replace [ansible-galaxy-extras](https://github.com/galaxyproject/ansible-galaxy-extras) with the roles that are maintained and e.g. used in [usegalaxy-eu/infrastructure-playbook](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/master/requirements.yaml) or on the [org server](https://github.com/galaxyproject/usegalaxy-playbook/blob/main/requirements.yml). Sorry that we did not came up with this earlier and you already updated the role in your fork. Maybe it makes the replacement easier, because you can replace the roles one by one and run the ci tests in between.
Sorry I think we are currently lacking a bit a use-case for the compose setup and we should have told you earlier.

jyotipm29 commented:
Thanks! I will check that out.
The current PR will track the single-container setup and [this](https://github.com/bgruening/docker-galaxy-stable/pull/608) will track the compose one.

jyotipm29 commented on .travis.yml:
```diff
@@ -39,6 +39,7 @@ before_install:
 
   # start building this repo
   - git submodule update --init --recursive
+  - git submodule update --remote --rebase
```
can be removed

jyotipm29 commented on .github/workflows/single.sh:
```diff
@@ -19,6 +19,7 @@ docker info
 
 # start building this repo
 git submodule update --init --recursive
+git submodule update --remote --rebase
```
can be removed

jyotipm29 commented on galaxy/Dockerfile:
```diff
@@ -2,7 +2,7 @@
 # VERSION       Galaxy-central
 
-FROM ubuntu:18.04
+FROM ubuntu:22.04
```
For this PR, I opted for Ubuntu 22.04 to align with HTCondor's current support, which covers Ubuntu 20.04 and 22.04 as per their [documentation](https://htcondor.readthedocs.io/en/latest/getting-htcondor/from-our-repositories.html).

bgruening replied:
That makes sense, thanks!

jyotipm29 commented on galaxy/tools_conf_interactive.xml.sample:
Yes it works, I tested a few tools like JupyterLab, Ethercalc, Rstudio etc.

bgruening replied:
Amazing!

jyotipm29 commented on galaxy/startup.sh:
```diff
@@ -23,47 +20,96 @@ if [ ! -f $GALAXY_CONFIG_FILE ]
+    export GALAXY_CONFIG_GALAXY_URL_PREFIX="$PROXY_PREFIX"
+    export GALAXY_CONFIG_INTERACTIVETOOLS_BASE_PATH="$PROXY_PREFIX"
+    
+    python3 /usr/local/bin/update_yaml_value "${GRAVITY_CONFIG_FILE}" "gravity.reports.url_prefix" "$PROXY_PREFIX/reports" &> /dev/null
```
I have added the python script [here](https://github.com/jyotipm29/ansible-galaxy-extras/blob/24.1/templates/update_yaml_value.py.j2).

sanjaysrikakulam commented:
> This looks really cool @jyotipm29 Thanks a lot! 🚀 From what I see, the tests are failing due to github walltime limits and storage limits. If you are sure that it works, maybe we could also test it manually again, we could already merge this. What do you think @sanjaysrikakulam ?
> 
> > Next, I’ll work on adding the Rustus service, integrating interactive tools, and replacing Nginx with Traefik in the compose setup. Let me know if you have any feedback on the current changes. Thanks!
> 
> A bit more important than the compose setup and traefik would be to successively replace [ansible-galaxy-extras](https://github.com/galaxyproject/ansible-galaxy-extras) with the roles that are maintained and e.g. used in [usegalaxy-eu/infrastructure-playbook](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/master/requirements.yaml) or on the [org server](https://github.com/galaxyproject/usegalaxy-playbook/blob/main/requirements.yml). Sorry that we did not came up with this earlier and you already updated the role in your fork. Maybe it makes the replacement easier, because you can replace the roles one by one and run the ci tests in between. Sorry I think we are currently lacking a bit a use-case for the compose setup and we should have told you earlier.

Sure! Björn already left his comments and suggestions, and I thought Jyoti would probably want to address them. However, as you pointed out, this can be merged.

@jyotipm29 Excellent work! Thank you! :)

bgruening commented:
@jyotipm29 what do you think about postponing the compose tests until after the single-container tests is green. This way we safe a bit of CI time and you can faster iterate on the single-container one?

bgruening commented on README.md:
```diff
@@ -350,20 +352,33 @@ docker exec -it <new_container_name> supervisorctl start galaxy:
-Interactive Environments (IE) are sophisticated ways to extend Galaxy with powerful services, like [Jupyter](http://jupyter.org/), in a secure and reproducible way.
+Interactive Tools (IT) are sophisticated ways to extend Galaxy with powerful services, like [Jupyter](http://jupyter.org/), in a secure and reproducible way.
 
 For this we need to be able to launch Docker containers inside our Galaxy Docker container. At least docker 1.3 is needed on the host system.
```
```suggestion
For this we need to be able to launch Docker containers inside our Galaxy Docker container.
```

jyotipm29 replied:
will update

jyotipm29 commented:
> @jyotipm29 what do you think about postponing the CI compose tests until after the single-container tests is green. This way we safe a bit of CI time and you can faster iterate on the single-container one?

Yes, good idea. I would temporarily disable those tests in the next commit.

jyotipm29 commented on README.md:
````diff
@@ -350,20 +352,33 @@ docker exec -it <new_container_name> supervisorctl start galaxy:
 For this we need to be able to launch Docker containers inside our Galaxy Docker container. At least docker 1.3 is needed on the host system.
 
 ```sh
-docker run -d -p 8080:80 -p 8021:21 -p 8800:8800 \
````
I am not sure why it would be needed

bgruening replied:
Isn't it the new proxy port, that you describe below?

jyotipm29 replied:
Oh yes, sorry, I will add that!

bgruening commented:
If someone wants to test it quickly :)

> docker run -p 8080:80 [quay.io/bgruening/galaxy:24.1-beta](http://quay.io/bgruening/galaxy:24.1-beta)

bgruening commented:
I get ...

> PermissionError: [Errno 13] Permission denied: '/home/galaxy/.config/conda/.condarc'

So installing tools into the container does not work with the container. Its strange, I thought we had a test for this.

jyotipm29 commented on galaxy/startup.sh:
```diff
@@ -23,47 +20,96 @@ if [ ! -f $GALAXY_CONFIG_FILE ]
+    $ANSIBLE_EXTRA_VARS_HTTPS_PROXY_PREFIX \
+    --tags https,$ANSIBLE_TAG_HTTPS_PROXY_PREFIX
+fi
+if [ "$USE_HTTPS" != "False" ]
```
![https](https://github.com/user-attachments/assets/231ece0d-9626-4005-b1e9-519cbaeffa18)
I tested SSL with letsencrypt and it works fine. The letsencrypt scripts needed some [updates](https://github.com/jyotipm29/ansible-galaxy-extras/commit/a1a98071796c069fb9b02920763f8abc3d763e60). However, But I used the dns-01 challenge to get the certs.

bgruening replied:
Cool, thanks for testing!

jyotipm29 commented:
> I get ...
> 
> > PermissionError: [Errno 13] Permission denied: '/home/galaxy/.config/conda/.condarc'
> 
> So installing tools into the container does not work with the container. Its strange, I thought we had a test for this.

This is weird. The tool installation worked in my environment.

bgruening commented:
Did you run with or without --privileged=true?

jyotipm29 commented:
worked both ways

bgruening commented:
Which tool.are you using to install?

jyotipm29 commented:
I tested [cherry_pick_fasta](https://toolshed.g2.bx.psu.edu/repository?repository_id=66756b6ac910d113) and [abyss](https://toolshed.g2.bx.psu.edu/repository?repository_id=82ab1904790830ef). Is there any particular tool that you want me to check?

jyotipm29 commented:
> This looks really cool @jyotipm29 Thanks a lot! 🚀 From what I see, the tests are failing due to github walltime limits and storage limits. If you are sure that it works, maybe we could also test it manually again, we could already merge this. What do you think @sanjaysrikakulam ?
> 
> > Next, I’ll work on adding the Rustus service, integrating interactive tools, and replacing Nginx with Traefik in the compose setup. Let me know if you have any feedback on the current changes. Thanks!
> 
> A bit more important than the compose setup and traefik would be to successively replace [ansible-galaxy-extras](https://github.com/galaxyproject/ansible-galaxy-extras) with the roles that are maintained and e.g. used in [usegalaxy-eu/infrastructure-playbook](https://github.com/usegalaxy-eu/infrastructure-playbook/blob/master/requirements.yaml) or on the [org server](https://github.com/galaxyproject/usegalaxy-playbook/blob/main/requirements.yml). Sorry that we did not came up with this earlier and you already updated the role in your fork. Maybe it makes the replacement easier, because you can replace the roles one by one and run the ci tests in between. Sorry I think we are currently lacking a bit a use-case for the compose setup and we should have told you earlier.

Just to confirm, the idea is to completely phase out ansible-galaxy-extras and instead use individual roles like usegalaxy_eu.nginx, usegalaxy_eu.htcondor etc, similar to how we currently use galaxyproject.postgresql in this repository. Is this correct?

bgruening commented:
> Just to confirm, the idea is to completely phase out ansible-galaxy-extras and instead use individual roles like usegalaxy_eu.nginx, usegalaxy_eu.htcondor etc, similar to how we currently use galaxyproject.postgresql in this repository. Is this correct?

Yes :-). And now since the tests work I would do that one commit at a time and see if tests still work.

bgruening commented on galaxy/Dockerfile:
```diff
@@ -86,91 +92,129 @@ RUN groupadd -r postgres -g $GALAXY_POSTGRES_GID \
+
+# Install ansible and related dependencies
+RUN apt update -qq && apt install --no-install-recommends -y software-properties-common dirmngr gpg gpg-agent \
+    && sudo add-apt-repository ppa:ansible/ansible \
```
Not sure which version of ansible you are installing here, but I get:

```
190.5 ERROR! [DEPRECATED]: ansible.builtin.include has been removed. Use include_tasks or import_tasks instead. This feature was removed from ansible-core in a release after 2023-05-16. Please update your playbooks.
190.5 
190.5 The error appears to be in '/ansible/roles/galaxyprojectdotorg.galaxyextras/tasks/main.yml': line 2, column 3, but may
190.5 be elsewhere in the file depending on the exact syntax problem.
190.5 
190.5 The offending line appears to be:
190.5 
190.5 
190.5 - include: galaxy_root.yml
190.5   ^ here
````

later below.

bgruening commented on galaxy/Dockerfile:
```diff
@@ -186,49 +230,54 @@ RUN mkdir -p /shed_tools $EXPORT_DIR/ftp/ \
     # Configure Galaxy to use the Tool Shed
     && chown $GALAXY_USER:$GALAXY_USER $EXPORT_DIR/tool_deps \
-    && apt update -qq && apt install --no-install-recommends -y ansible \
+    && apt update -qq && apt install --no-install-recommends -y ansible gpg gpg-agent g++ make bsdmainutils \
```
Here ansible is installed again.

bgruening replied:
```suggestion
    && apt update -qq && apt install --no-install-recommends -y gpg gpg-agent g++ make bsdmainutils \
```

We should not need g++, which is something to investigate at the very end.

jyotipm29 replied:
I would check that.

jyotipm29 replied:
I have removed g++

jyotipm29 commented:
While installing cvmfs client from [ansible-cvmfs](https://github.com/galaxyproject/ansible-cvmfs) role during docker build, it requires autofs to be [running](https://github.com/galaxyproject/ansible-cvmfs/blob/950aed64817cd104715bdb67869cbbfca84e1111/tasks/client.yml#L17), and autofs can't run without privileged mode. Do you have any suggestions for this?

sanjaysrikakulam commented:
> While installing cvmfs client from [ansible-cvmfs](https://github.com/galaxyproject/ansible-cvmfs) role during docker build, it requires autofs to be [running](https://github.com/galaxyproject/ansible-cvmfs/blob/950aed64817cd104715bdb67869cbbfca84e1111/tasks/client.yml#L17), and autofs can't run without privileged mode. Do you have any suggestions for this?

Maybe this might help: https://github.com/cvmfs/cvmfsexec

bgruening commented:
See https://github.com/bgruening/docker-galaxy-stable/issues/609 for a few more information.

jyotipm29 commented on galaxy/Dockerfile:
```diff
@@ -48,220 +49,217 @@ ENV GALAXY_CONFIG_FILE=$GALAXY_CONFIG_DIR/galaxy.yml \
     && deactivate \
-    # TODO: no clue why this is needed here again
-    && cd $GALAXY_ROOT && ./scripts/common_startup.sh \
+    # this is needed here again to install any conditional dependencies
```
These 2 packages get installed by this step: psycopg2-binary==2.9.9 and drmaa

bgruening replied:
I guess its coming from https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/conditional-requirements.txt  so we can simply call this and remove the common_startup for a second time.

jyotipm29 replied:
I will add this

jyotipm29 replied:
We could set [LOAD_GALAXY_CONDITIONAL_DEPENDENCIES](https://github.com/bgruening/docker-galaxy-stable/blob/42e298e95e3c1064b5f13e9467a0f388aed569c5/galaxy/startup.sh#L123) and it would install the necessary packages during startup.

bgruening replied:
Nice, lets do that.

bgruening commented:
@jyotipm29 have you ever seen this error?

> TASK [Run common_startup.sh to build the galaxy client] ************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "scripts/common_startup.sh", "delta": "0:03:36.364695", "end": "2024-11-07 10:52:57.276437", "msg": "non-zero return code", "rc": 1, "start": "2024-11-07 10:49:20.911742", "stderr": "ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: 'METADATA'\nCheck the permissions."

jyotipm29 commented:
> @jyotipm29 have you ever seen this error?
> 
> > TASK [Run common_startup.sh to build the galaxy client] ************************
> > fatal: [localhost]: FAILED! => {"changed": true, "cmd": "scripts/common_startup.sh", "delta": "0:03:36.364695", "end": "2024-11-07 10:52:57.276437", "msg": "non-zero return code", "rc": 1, "start": "2024-11-07 10:49:20.911742", "stderr": "ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: 'METADATA'\nCheck the permissions."

No, I haven't seen this

bgruening commented on galaxy/Dockerfile:
```diff
@@ -176,12 +176,11 @@ RUN mkdir -p /shed_tools $GALAXY_CONFIG_TUS_UPLOAD_STORE \
     && ln -s /tool_deps/ $EXPORT_DIR/tool_deps \
     # Configure Galaxy to use the Tool Shed
-    && chown $GALAXY_USER:$GALAXY_USER $EXPORT_DIR/tool_deps \
+    && chown $GALAXY_USER:$GALAXY_USER $EXPORT_DIR/tool_deps $GALAXY_VIRTUAL_ENV -R \
```
The venv had files owned by root and that caused problems for me. I have no idea why this only happens for me and not for you and the CI.

jyotipm29 replied:
The flower installation just before this task installs packages into that directory. I think that was the issue. I hadn’t built the image locally since the last few refactorings, but now, when I do, I see the problem. Surprisingly, the build was successful in the CI environment. Thanks for fixing it!

bgruening replied:
Can we convince the step before to run the pip install step as `galaxy` user?

jyotipm29 replied:
There is no option at the [task](https://github.com/usegalaxy-eu/flower-ansible-role/blob/19e75b0876d59713fcd339064bbbf33d4716e7e3/tasks/main.yml#L58) level. Since some other tasks require root privileges, I can't set the user at play level.

bgruening commented:
It seems that if you start Galaxy with `-v foo:/export`, close it and reopen it does not start again (nginx error). This is an important functionality, as it allows persiting the state of the application.

jyotipm29 commented:
I can't reproduce the issue. I have tested it and the persistence works. Please share what error you see.

bgruening commented:
Start the contianer, install a tool.
Look at `(base) root@b1b310dd2cfc:/galaxy-central# ll /export/shed_tools/` there is nothing under /export.

It is here:

```
(base) root@b1b310dd2cfc:/galaxy-central# head /etc/galaxy/shed_tool_conf.xml 
<?xml version="1.0" ?>
<toolbox tool_path="/galaxy-central/database/shed_tools">
    <tool file="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/541082d03bef/samtools_stats/samtools_stats.xml" guid="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5">
        <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
        <repository_name>samtools_stats</repository_name>
        <repository_owner>devteam</repository_owner>
        <installed_changeset_revision>541082d03bef</installed_changeset_revision>
        <id>toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5</id>
        <version>2.0.5</version>
    </tool>
(base) root@b1b310dd2cfc:/galaxy-central# ll /galaxy-central/database/shed_tools/
total 0
drwxr-xr-x 3 galaxy galaxy 60 Nov  8 10:10 toolshed.g2.bx.psu.edu/
(base) root@b1b310dd2cfc:/galaxy-central# 
```

When you now stop the container and run the same command again, so mount the same directory in again. It can not load the installed tool, because it was not in the /export dir. the result is that the tool is not installed anymore. But there is also a different problem. I only see an nginx error page. Galaxy is not starting in this setting.

jyotipm29 commented:
> Start the contianer, install a tool. Look at `(base) root@b1b310dd2cfc:/galaxy-central# ll /export/shed_tools/` there is nothing under /export.
> 
> It is here:
> 
> ```
> (base) root@b1b310dd2cfc:/galaxy-central# head /etc/galaxy/shed_tool_conf.xml 
> <?xml version="1.0" ?>
> <toolbox tool_path="/galaxy-central/database/shed_tools">
>     <tool file="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/541082d03bef/samtools_stats/samtools_stats.xml" guid="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5">
>         <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
>         <repository_name>samtools_stats</repository_name>
>         <repository_owner>devteam</repository_owner>
>         <installed_changeset_revision>541082d03bef</installed_changeset_revision>
>         <id>toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5</id>
>         <version>2.0.5</version>
>     </tool>
> (base) root@b1b310dd2cfc:/galaxy-central# ll /galaxy-central/database/shed_tools/
> total 0
> drwxr-xr-x 3 galaxy galaxy 60 Nov  8 10:10 toolshed.g2.bx.psu.edu/
> (base) root@b1b310dd2cfc:/galaxy-central# 
> ```
> 
> When you now stop the container and run the same command again, so mount the same directory in again. It can not load the installed tool, because it was not in the /export dir. the result is that the tool is not installed anymore. But there is also a different problem. I only see an nginx error page. Galaxy is not starting in this setting.

Since we already persist the database folder, do we need the /export/shed_tools dir?

bgruening commented on galaxy/ansible/templates/export_user_files.py.j2:
```diff
@@ -142,7 +142,7 @@ if __name__ == "__main__":
     change_path( os.path.join(galaxy_root_dir, 'database') )
-    change_path( '/shed_tools/' )
+    change_path( '/tus_upload_store/' )
 
```
Tus upload store are temporary files during the upload. I don't think we need to persist them.

jyotipm29 replied:
will remove it

bgruening commented on galaxy/reports.yml.sample:
```diff
@@ -11,10 +11,10 @@ reports:
 
   # Where temporary files are stored.
-  #new_file_path: database/tmp
+  new_file_path: database/files
```
Why this?

jyotipm29 replied:
I thought of keep it consistent with galaxy [config](https://github.com/bgruening/docker-galaxy-stable/blob/42e298e95e3c1064b5f13e9467a0f388aed569c5/galaxy/Dockerfile#L247). I will revert it back.

bgruening replied:
Upstream is setting it to https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/galaxy.yml.sample#L495  so I guess we should change the ENV var instead.

bgruening commented on galaxy/Dockerfile:
```diff
@@ -110,77 +48,106 @@ RUN curl -s -L https://github.com/conda-forge/miniforge/releases/download/$MINIF
+ADD ./common_cleanup.sh /usr/bin/common_cleanup.sh
+RUN chmod +x /usr/bin/common_cleanup.sh
+
+RUN echo "force-unsafe-io" > /etc/dpkg/dpkg.cfg.d/02apt-speedup \
```
If it's not the final container we do not need to cleanup isn't it? That is one of the advantages of a multi stage build. It makes the file more readable.

jyotipm29 replied:
Yes true. I will do the changes.

jyotipm29 commented:
> Start the contianer, install a tool. Look at `(base) root@b1b310dd2cfc:/galaxy-central# ll /export/shed_tools/` there is nothing under /export.
> 
> It is here:
> 
> ```
> (base) root@b1b310dd2cfc:/galaxy-central# head /etc/galaxy/shed_tool_conf.xml 
> <?xml version="1.0" ?>
> <toolbox tool_path="/galaxy-central/database/shed_tools">
>     <tool file="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/541082d03bef/samtools_stats/samtools_stats.xml" guid="toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5">
>         <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
>         <repository_name>samtools_stats</repository_name>
>         <repository_owner>devteam</repository_owner>
>         <installed_changeset_revision>541082d03bef</installed_changeset_revision>
>         <id>toolshed.g2.bx.psu.edu/repos/devteam/samtools_stats/samtools_stats/2.0.5</id>
>         <version>2.0.5</version>
>     </tool>
> (base) root@b1b310dd2cfc:/galaxy-central# ll /galaxy-central/database/shed_tools/
> total 0
> drwxr-xr-x 3 galaxy galaxy 60 Nov  8 10:10 toolshed.g2.bx.psu.edu/
> (base) root@b1b310dd2cfc:/galaxy-central# 
> ```
> 
> When you now stop the container and run the same command again, so mount the same directory in again. It can not load the installed tool, because it was not in the /export dir. the result is that the tool is not installed anymore. But there is also a different problem. I only see an nginx error page. Galaxy is not starting in this setting.

Setting [managed_config_dir](https://github.com/galaxyproject/galaxy/blob/00022427d5b83a0e3172b8bf24743d9fae334276/lib/galaxy/config/sample/galaxy.yml.sample#L120) to `database/config` would fix this i guess (since `database` is persisted and earlier the config pointed to `/etc/galaxy` by default)

bgruening commented:
I will document what I found so far here, not everything needs to be addressed, its just a list to keep track.

I'm tempted to merge as it is and close old issues, ping old users to give this new version a try as soon as we can ensure the /export functionality works.

* Major, we need to ensure that most, all useful tests are running. See my comments about the old travis stuff. New tests can also not hurt.

* I was still not able to get the container running with an exported version of data under `/export`. My Galaxy does simply not start under those conditions, even if I wait for multiple minutes.

When I restart a container with a full `/export`:
```
(base) root@fc28a50a6928:/galaxy-central# galaxyctl status
Dynamic handlers are configured in Gravity but Galaxy is not configured to assign jobs to handlers dynamically, so these handlers will not handle jobs. Set the job handler assignment method in the Galaxy job configuration to `db-skip-locked` or `db-transaction-isolation` to fix this.
supervisord is not running
```

* [2024-11-09T16:40:51.456] error: Node configuration differs from hardware: CPUs=12:12(hw) Boards=1:1(hw) SocketsPerBoard=12:1(hw) CoresPerSocket=1:6(hw) ThreadsPerCore=1:2(hw)


Cosmetic changes:
* /root/502/index.shtml does not exist - it is a configured nginx error page
* the image is still double in size of the last version [`20.09`](https://quay.io/repository/bgruening/galaxy) but this we can investigate another time
* you did a great job in adding `galaxy-root-path` everywhere, maybe its time to rename the entire folder just to `galaxy` and not `galaxy-central`. This is really a old name no one remembers anymore :)

bgruening commented:
```
(base) root@fc28a50a6928:/galaxy-central# ll -la /export/
total 4
drwxr-xr-x 8 root     root      160 Nov  9 16:40 ./
drwxr-xr-x 1 root     root     4096 Nov  9 16:40 ../
drwxrwxr-x 3 root     root      720 Nov  8 08:22 .distribution_config/
drwxr-xr-x 2 galaxy   galaxy     40 Nov  9 16:36 ftp/
drwxr-xr-x 7 galaxy   galaxy    180 Nov  9 16:40 galaxy-central/
drwxr-xr-x 3 postgres postgres   60 Nov  9 16:36 postgresql/
drwxr-xr-x 3 galaxy   galaxy     80 Nov  9 16:38 tool_deps/
drwxr-xr-x 3 root     root       60 Nov  9 16:36 var/
```

```
(base) root@fc28a50a6928:/galaxy-central# ll -la /export/.distribution_config/
total 528
drwxrwxr-x 3 root root    720 Nov  8 08:22 ./
drwxr-xr-x 8 root root    160 Nov  9 16:40 ../
-rw-rw-r-- 1 root root   6736 Nov  8 08:22 auth_conf.xml.sample
-rw-rw-r-- 1 root root   1080 Nov  8 08:22 build_sites.yml.sample
-rw-rw-r-- 1 root root   4761 Nov  8 08:22 container_resolvers.yml.sample
-rw-rw-r-- 1 root root     55 Nov  8 08:22 data_manager_conf.xml.sample
-rw-rw-r-- 1 root root 122919 Nov  8 08:22 datatypes_conf.xml.sample
-rw-rw-r-- 1 root root    412 Nov  8 08:22 disposable_email_blocklist.conf.sample
-rw-rw-r-- 1 root root   2555 Nov  8 08:22 environment_modules_mapping.yml.sample
-rw-rw-r-- 1 root root   4555 Nov  8 08:22 error_report.yml.sample
-rw-rw-r-- 1 root root   8852 Nov  8 08:22 file_sources_conf.yml.sample
-rw-rw-r-- 1 root root   1078 Nov  8 08:22 galaxy.yml.interactivetools
-rw-rw-r-- 1 root root 137140 Nov  8 08:22 galaxy.yml.sample
-rw-rw-r-- 1 root root  53508 Nov  8 08:22 job_conf.sample.yml
-rw-rw-r-- 1 root root   2125 Nov  8 08:22 job_conf.yml.interactivetools
-rw-rw-r-- 1 root root   1050 Nov  8 08:22 job_conf.yml.interactivetools.podman
-rw-rw-r-- 1 root root    698 Nov  8 08:22 job_resource_params_conf.xml.sample
-rw-rw-r-- 1 root root   1950 Nov  8 08:22 lmod_modules_mapping.yml.sample
-rw-rw-r-- 1 root root    262 Nov  8 08:22 local_conda_mapping.yml.sample
-rw-rw-r-- 1 root root  22698 Nov  8 08:22 object_store_conf.sample.yml
-rw-rw-r-- 1 root root  13100 Nov  8 08:22 oidc_backends_config.xml.sample
-rw-rw-r-- 1 root root   1199 Nov  8 08:22 oidc_config.xml.sample
drwxrwxr-x 5 root root    100 Nov  8 08:22 plugins/
-rw-rw-r-- 1 root root   1390 Nov  8 08:22 reports.yml.sample
-rw-rw-r-- 1 root root   1594 Nov  8 08:22 themes_conf.yml.sample
-rw-rw-r-- 1 root root   7130 Nov  8 08:22 tool_conf.xml.sample
-rw-rw-r-- 1 root root   8305 Nov  8 08:22 tool_data_table_conf.xml.sample
-rw-rw-r-- 1 root root   9732 Nov  8 08:22 tool_destinations.yml.sample
-rw-rw-r-- 1 root root   1790 Nov  8 08:22 tool_recommendations_overwrite.yml.sample
-rw-rw-r-- 1 root root    288 Nov  8 08:22 tool_sheds_conf.xml.sample
-rw-rw-r-- 1 root root  18111 Nov  8 08:22 tool_shed.yml.sample
-rw-rw-r-- 1 root root    784 Nov  8 08:22 trs_servers_conf.yml.sample
-rw-rw-r-- 1 root root   6227 Nov  8 08:22 user_preferences_extra_conf.yml.sample
-rw-rw-r-- 1 root root    334 Nov  8 08:22 workflow_resource_mapper_conf.yml.sample
-rw-rw-r-- 1 root root   1136 Nov  8 08:22 workflow_resource_params_conf.xml.sample
-rw-rw-r-- 1 root root   1101 Nov  8 08:22 workflow_schedulers_conf.xml.sample
```

330 MB ... what is that? A leftover from ansible?

bgruening commented on README.md:
````diff
@@ -880,23 +916,30 @@ If you want to create new users, please make sure to use the `/export/` volume.
+You can clone this repository with:
 
 ```sh
 git clone --recursive https://github.com/bgruening/docker-galaxy-stable.git
````
```suggestion
git clone https://github.com/bgruening/docker-galaxy-stable.git
```

So cool, that we can get rid of the submodules now :)

bgruening commented on galaxy/common_cleanup.sh:
```diff
@@ -0,0 +1,19 @@
+
+find / -name '*.log' -delete
+find / -name '.cache' -delete
+find / -name '.npm' -delete
```
I spotted a few folders called `.parcel-cache` under galaxy/config/plugins with significant size, maybe that can be removed as well

bgruening replied:
/root/.launchpadlib is 2MB and added in the "apt-get -qq update     && apt install -y software-properties-common dirmngr gpg gpg-agent   " step

jyotipm29 replied:
will remove these...thanks for noticing!

bgruening commented:
```
(base) root@cdb56fab569a:/galaxy-central# ll -ah /home/galaxy/
total 60K
drwxr-x--- 1 galaxy galaxy 4.0K Nov  9 18:30 ./
drwxr-xr-x 1 root   root   4.0K Nov  9 15:57 ../
drwxr-xr-x 3 root   root   4.0K Nov  9 18:02 .ansible/
-rw------- 1 galaxy galaxy   39 Nov  9 18:30 .bash_history
-rw-r--r-- 1 galaxy galaxy  220 Jan  6  2022 .bash_logout
-rw-r--r-- 1 galaxy galaxy 3.4K Nov  9 14:46 .bashrc
drwxrwxr-x 1 galaxy galaxy 4.0K Nov  9 18:03 .cache/
drwx------ 3 root   root   4.0K Nov  9 18:10 .config/
drwxr-xr-x 1 root   root   4.0K Nov  9 16:21 ephemeris/
```
Some folders are owned by root, this means that certain operations are crashing under the galaxy user.

jyotipm29 commented on .travis.yml:
I can do that

jyotipm29 commented:
> ```
> (base) root@cdb56fab569a:/galaxy-central# ll -ah /home/galaxy/
> total 60K
> drwxr-x--- 1 galaxy galaxy 4.0K Nov  9 18:30 ./
> drwxr-xr-x 1 root   root   4.0K Nov  9 15:57 ../
> drwxr-xr-x 3 root   root   4.0K Nov  9 18:02 .ansible/
> -rw------- 1 galaxy galaxy   39 Nov  9 18:30 .bash_history
> -rw-r--r-- 1 galaxy galaxy  220 Jan  6  2022 .bash_logout
> -rw-r--r-- 1 galaxy galaxy 3.4K Nov  9 14:46 .bashrc
> drwxrwxr-x 1 galaxy galaxy 4.0K Nov  9 18:03 .cache/
> drwx------ 3 root   root   4.0K Nov  9 18:10 .config/
> drwxr-xr-x 1 root   root   4.0K Nov  9 16:21 ephemeris/
> ```
> 
> Some folders are owned by root, this means that certain operations are crashing under the galaxy user.

The root user's [home](https://github.com/jyotipm29/docker-galaxy-stable/blob/42e298e95e3c1064b5f13e9467a0f388aed569c5/galaxy/Dockerfile#L266) directory was set to /home/galaxy. I will fix it in the next commit.

sanjaysrikakulam reviewed (approved):
I think this is ready to be merged. I ran a quick test, and I was able to spin it up, install the tools, upload files, run jobs, and run ITs. 

Great work @jyotipm29!

jyotipm29 commented:
> I think this is ready to be merged. I ran a quick test, and I was able to spin it up, install the tools, upload files, run jobs, and run ITs.
> 
> Great work @jyotipm29!

Thanks! I really appreciate it! :)

bgruening commented:
Great work!
